---
title: "MicroEcoTools"
author: "Soheil A. Neshat"
date: "2025-03-10"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{MicroEcoTools Vignette}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  markdown: 
    wrap: 72
---

```{r setup, include=FALSE}
# Set global options for code chunks
knitr::opts_chunk$set(echo = TRUE, warning = TRUE, message = TRUE)
```

## Introduction

MicroEcoTools is an R package that empowers microbial ecologists with a comprehensive suite of computational tools to apply theoretical frameworks to community data. It implements methods for testing hypotheses on communities under disturbance, including both trait-based approaches—using Grime's Competitive–Stress tolerant–Ruderal (CSR) framework—and null model analyses (NMA). The package supports the analysis of complex datasets derived from perturbation experiments and other ecological studies. The manuscript is available through BioRxiV: "Neshat, Soheil A., Ezequiel Santillan, and Stefan Wuertz. "MicroEcoTools: an R package for comprehensive theoretical microbial ecology analysis." bioRxiv (2024): 2024-08."

**Key functionalities include:**
- **Pairwise Comparisons and CSR Assignment:** Automatically assign CSR categories to taxa or traits based on pairwise statistical comparisons.
- **Simulation-based CSR Assignment:** Generate simulated community data using a multinomial distribution to assess the probability of CSR assignments.
- **Null model analysis:** Calculate the relative contribution of stochastic and deterministic assembly mechanisms.

This vignette provides a step-by-step guide to run the core analyses offered by the package. The tools within MicroEcoTools enable users to investigate the relative contributions of deterministic and stochastic assembly mechanisms in microbial communities and to assign Grime's CSR categories under varying disturbance regimes. We use data sets originally published in the following publications:
1.Santillan, Ezequiel, et al. "Frequency of disturbance alters diversity, function, and underlying assembly mechanisms of complex bacterial communities." npj Biofilms and Microbiomes 5.1 (2019): 8.
2. Santillan, Ezequiel, et al. "Trait‐based life‐history strategies explain succession scenario for complex bacterial communities under varying disturbance." Environmental microbiology 21.10 (2019): 3751-3764.

## Installation
Running the following commands will install the MicroEcoTools package:
```{r installation}
# Check if the MicroEcoTools package is installed and if its version is at least 0.4
  if (requireNamespace("MicroEcoTools", quietly = TRUE) && packageVersion("MicroEcoTools") < "0.4") {
    remove.packages("MicroEcoTools")
  }

if (!requireNamespace("MicroEcoTools", quietly = TRUE)) {
  # If not installed or version is lower than 0.4, install from GitHub
  devtools::install_github("Soheil-A-Neshat/MicroEcoTools")
}

# Load the MicroEcoTools package
library(MicroEcoTools)
```

##Research questions and experimental design
The following text is borrowed from references 1 and 2 with some modifications:
"The objective of the original research work was to test the effect of disturbance on the bacterial community structure, diversity, and ecosystem function of a complex bacterial system, with emphasis on the underlying assembly mechanisms. The researchers employed sequencing batch bioreactors inoculated with activated sludge from an urban wastewater treatment plant, in a laboratory microcosm setup with eight different frequency levels of augmentation with toxic 3-chloroaniline (3-CA) as disturbance. Triplicate reactors received 3- CA either never (L0, undisturbed), every 7, 6, 5, 4, 3, and 2 days (L1–6, intermediately-disturbed), or every day (L7, press-disturbed) for 35 days. Chloroanilines are toxic and carcinogenic compounds and few bacteria encode the pathways to degrade 3-CA,41 which is also known to inhibit both organic carbon removal and nitrification in sludge reactors. The authros also we investigated the effect of disturbance on bacte-rial community structure, genotype and function within a framework of three-way CSR life-history strategies."

In this tutorial, the focus is on applying Grime's CSR framework and assessing the assembly mechanisms. We strongly recommend to read the original works if you are intrested in the topic.

## Data
The data generated in the mentioned studies are included in MicroEcoTools and can be imported into the environment by running the following chunk:
```{r data import}
# Importing the data into the R environment
data(NMA_data)
data(CSR_IP2G_data)
```

It's generally recommended to investigate any data sets that you intend to work on before running any analysis. Running the following chunk shows a subset and structure of the data:
```{r data investigation - NMA_data}
# Investigating the structure of the data
str(NMA_data)
dim(NMA_data)
head(NMA_data)
```

```{r data investigation - CSR_IP2G_data}
# Investigating the structure of the data
str(CSR_IP2G_data)
dim(CSR_IP2G_data)
head(CSR_IP2G_data)
```

It's recomended to use the included datasets' format as a guide and format your data sets similarly before using MicroEcoTools.

If your data are stored as a SummarizedExperiment or phyloseq object, the functions automatically check the input and call the helper function ?convert_to_microecotools_df to convert it into the required format. Please note that this function is currently released as a beta version and errors might occur under certain conditions.


Now that we have investigated the data sets, we can proceed with the analysis.
First we will apply the Grime's CSR framework on the CSR_IP2G_data to assign competitive, stress-tolerant, and ruderal categories to functions.

##CSR Assignment
CSR_assign function analyses the dataset and assigns competitive, stress-tolerant and ruderal categories to traits/taxa which are abundant in undisturbed, highly disturbed and intermediately disturbed communities. The output for this function is a list containing comparison tables and required materials for the plotting function (CSR_plot).
CSR_assign function requires a data set to perform the analysis. The rest of the arguments have set defaults to conduct the analysis. You can investigate the parameters in the help function and change the parameters if required.

```{r CSR assignment}
# Perform CSR category assignment
csr_result <- CSR_assign(dAtA = CSR_IP2G_data)
#csr_result <- CSR_assign(dAtA = CSR_IP2G_data, var.name = "Trait", p_adj = "BH", v.equal = FALSE, p.value.cutoff = 0.05, Parallel = TRUE, Verbose = FALSE)
```
To view the summary output of the assignment:
```{r CSR assignment summary, echo=FALSE}
# Investigating the summary of the CSR category assignment
summary(csr_result)
```

Then, we can visualise the CSR assignments:
```{r CSR plotting, echo=FALSE}
# Plot the traits along with their assigned CSR categories:
# Traits sorted based on abundance 
CSR_plot(CSR_results = csr_result, plot_type = "abundance")
# Traits sorted based on p-value
# CSR_plot(CSR_results = csr_result, plot_type = "p_value")
```


##CSR Assignment with simulation 
CSR_simulation function generates simulated communities with similar multinomial distribution to the observed communities and performs CSR category assignment on observed and simulated communities. The difference between this function and CSR_assign is that this function calculates a probability value which shows accuracy of the category assignment.

This function requires an input data frame along with the number of simulations. We recommend using at least 1000 simulations to achieve more accurate analysis. Here, we used 3 simulations to just demonstrate the functionality:
```{r CSR assignment with simulation}
# Run CSR_Simulation on the dataset.
#simulation_result <- CSR_Simulation(DaTa = CSR_IP2G_data, NSim = 2)
simulation_result <- CSR_Simulation(DaTa = CSR_IP2G_data, NSim = 2, p_adj = "BH", var.name = "Trait", v.equal = FALSE, p.value.cutoff = 0.05, Parallel = TRUE, NuLl.test = FALSE, Keep_data = FALSE)
```
We can check the output table by:
```{r CSR assignment with simulation output inspection, echo=FALSE}
View(simulation_result$Final_Verdict)
summary(simulation_result)
```
##Null model analysis 
To assess the relative contribution of stochastic and deterministic assembly mechanisms in a community using the NMA function. Running the following code chunk does this analysis. This function takes in the observed communities, pool all the taxa and generate random communities from the pool. Then, it compares the observed communities and randomly generated communities to calculate an effect size. The higher the effect size the more deterministic the community assembly.

```{r Null model analysis}
nma_results <- NMA(DaTa = NMA_data, NSim = 10, InDeX = "invsimpson")
summary(nma_results)
```

```{r Null model analysis visualisation, echo=FALSE}
#you can look at the standard effect size plot generated using richness as the measure of diversity using
NMA_plot(nma_results, plot_type = "SES_species_number")
#To see the plot generated for standard effect size using the specified diversity index use
NMA_plot(nma_results, plot_type = "SES")
```
The lower the standard effect size the greater contribution of stochastic assembly mechanisms.

## Conclusion

In this vignette, we demonstrated how MicroEcoTools can be used to analyze microbial community data through both null model analysis and trait‐based CSR assignments. By walking through the data preparation, CSR assignment, and null model analysis workflows, we illustrated the package's flexibility in handling complex experimental designs and its ability to generate intuitive visual summaries of microbial assembly mechanisms. We encourage users to explore the package with their own datasets and adapt the provided workflows to their research questions. 